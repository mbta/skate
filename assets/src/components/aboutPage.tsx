import React, { ReactElement } from "react"

const AboutPage = (): ReactElement<HTMLDivElement> => (
  <div className="c-page c-page--about">
    <div className="c-page__container">
      <h1 className="c-page__title">About Skate</h1>
      <p>
        Skate is an active work-in-progress. In the coming months, you’ll notice
        new tools like shuttle tracking. You can also expect changes and
        refinements to the route ladders as we continue building.
      </p>
      <div className="c-page__section">
        <h2 className="c-page__header">Device Support</h2>
        <p>Need tablet support? Call the MBTA Help Desk at x5761.</p>
      </div>
      <div className="c-page__section">
        <h2 className="c-page__header">Frequently Asked Questions</h2>
        <p>
          Have questions about using Skate? We’re making updates all the time.
          Read the latest news and refresh your training by visiting our FAQ.
        </p>
        <p>
          <a
            className="c-page__link"
            href="https://docs.google.com/document/d/e/2PACX-1vQVzXUcQWg-nKCkb8CyModNUCXDHHRxgqPdtlkm1GQOK9aCrGMSJAoC2J8JuuGFPGnPkcetrZEk6ERH/pub"
            target="_blank"
          >
            FAQ
          </a>
        </p>
      </div>
      <div className="c-page__section">
        <h2 className="c-page__header">Feedback</h2>
        <p>
          Have feedback, a suggestion, or need other support related to Skate?
          Use our feedback form and someone on our team will be in touch with
          you.
        </p>
        <p>
          <a
            className="c-page__link"
            href="https://form.asana.com?hash=e0ca8ee9a5fbde7691ba0ff428c826eba5a9700b99249e26124dd33d13b511ca&id=1155343357379913"
            target="_blank"
          >
            Give Feedback
          </a>
        </p>
      </div>
    </div>
  </div>
)

export default AboutPage
