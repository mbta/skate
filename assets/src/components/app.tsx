import React, { ReactElement, useContext } from "react"
import { BrowserRouter, Route as BrowserRoute } from "react-router-dom"
import { SocketContext } from "../contexts/socketContext"
import { StateDispatchContext } from "../contexts/stateDispatchContext"
import useAppcues from "../hooks/useAppcues"
import { ConnectionStatus } from "../hooks/useSocket"
import DataStatusBanner from "./dataStatusBanner"
import DisconnectedModal from "./disconnectedModal"
import LadderPage from "./ladderPage"
import SearchPage from "./searchPage"
import SettingsPage from "./settingsPage"
import ShuttleMapPage from "./shuttleMapPage"
import TabBar from "./tabBar"
import { Notifications } from "./notifications"

const AppRoutes = () => {
  useAppcues()

  const [{ pickerContainerIsVisible }] = useContext(StateDispatchContext)
  const { connectionStatus } = useContext(SocketContext)

  return (
    <div className="m-app">
      <div className="m-app__banner">
        <DataStatusBanner />
      </div>
      <div className="m-app__main">
        <Notifications />
        <TabBar pickerContainerIsVisible={pickerContainerIsVisible} />
        <BrowserRoute exact={true} path="/" component={LadderPage} />
        <BrowserRoute
          exact={true}
          path="/shuttle-map"
          component={ShuttleMapPage}
        />
        <BrowserRoute exact={true} path="/settings" component={SettingsPage} />
        <BrowserRoute exact={true} path="/search" component={SearchPage} />
        {connectionStatus === ConnectionStatus.Disconnected ? (
          <DisconnectedModal />
        ) : null}
      </div>
    </div>
  )
}

const App = (): ReactElement<HTMLDivElement> => {
  return (
    <BrowserRouter>
      <AppRoutes />
    </BrowserRouter>
  )
}

export default App
