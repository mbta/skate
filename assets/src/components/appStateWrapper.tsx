import React from "react"
import { StateDispatchProvider } from "../contexts/stateDispatchContext"
import { TripsByIdProvider } from "../contexts/tripsByIdContext"
import { VehiclesByRouteIdProvider } from "../contexts/vehiclesByRouteIdContext"
import usePersistedStateReducer from "../hooks/usePersistedStateReducer"
import useSocket from "../hooks/useSocket"
import useTrips from "../hooks/useTrips"
import useVehicles from "../hooks/useVehicles"
import { TripsById } from "../schedule.d"
import { initialState, reducer } from "../state"
import App from "./app"

const AppStateWrapper = (): JSX.Element => {
  const [state, dispatch] = usePersistedStateReducer(reducer, initialState)
  const { selectedRouteIds } = state

  const tripsById: TripsById = useTrips(selectedRouteIds)
  const socket = useSocket()
  const vehiclesByRouteId = useVehicles(socket, selectedRouteIds)

  return (
    <StateDispatchProvider state={state} dispatch={dispatch}>
      <TripsByIdProvider tripsById={tripsById}>
        <VehiclesByRouteIdProvider vehiclesByRouteId={vehiclesByRouteId}>
          <App />
        </VehiclesByRouteIdProvider>
      </TripsByIdProvider>
    </StateDispatchProvider>
  )
}

export default AppStateWrapper
