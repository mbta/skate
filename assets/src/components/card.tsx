import React from "react"
import { formattedTimeDiffUnderThreshold } from "../util/dateTime"
import { unreadIcon } from "../helpers/icon"
import CloseButton from "./closeButton"

export type CardStyle = "kiwi" | "white"

interface CardProps {
  style: CardStyle
  currentTime?: Date
  openCallback?: () => void
  closeCallback?: () => void
  isUnread?: boolean
  additionalClass?: string
  title: string
  time?: Date
  noFocusOrHover?: boolean
}

export const Card: React.FC<CardProps> = ({
  children,
  style,
  currentTime,
  openCallback,
  closeCallback,
  isUnread,
  additionalClass,
  title,
  time,
  noFocusOrHover,
}) => {
  const innerLeftContent = (
    <>
      <div className="m-card__top-row">
        <div className="m-card__title">
          {isUnread ? unreadIcon() : null}
          {title}
        </div>
        {currentTime && time ? (
          <div className="m-card__time">
            {formattedTimeDiffUnderThreshold(currentTime, time, 60)}
          </div>
        ) : null}
      </div>
      <div className="m-card__contents">{children}</div>
    </>
  )

  return (
    <div
      className={
        `m-card m-card--${style}` +
        (additionalClass ? " " + additionalClass : "") +
        (noFocusOrHover ? " m-card--no-focus-or-hover" : "") +
        (!isUnread ? " m-card--read" : "")
      }
    >
      {openCallback ? (
        <button className="m-card__left" onClick={openCallback}>
          {innerLeftContent}
        </button>
      ) : (
        <div className="m-card__left">{innerLeftContent}</div>
      )}
      {closeCallback ? (
        <div className="m-card__right">
          <CloseButton closeButtonType="xl_green" onClick={closeCallback} />
        </div>
      ) : null}
    </div>
  )
}

export const CardBody: React.FC = ({ children }) => (
  <div className="m-card__body">{children}</div>
)

export interface Property {
  label: string
  value: string | null
  sensitive?: boolean
}

export interface CardPropertiesProps {
  properties: Property[]
}

export const CardProperties: React.VFC<CardPropertiesProps> = ({
  properties,
}: CardPropertiesProps) => {
  return (
    <ul className="m-card__properties">
      {properties.map((property) =>
        property.value ? (
          <li key={property.label}>
            <span className="m-card__properties-label">{property.label}</span>
            <span
              className={
                "m-card__properties-value" +
                (property.sensitive
                  ? " m-card__properties-value--sensitive"
                  : "")
              }
            >
              {property.value}
            </span>
          </li>
        ) : null
      )}
    </ul>
  )
}
