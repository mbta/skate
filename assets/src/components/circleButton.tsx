import React, { ReactNode } from "react"
import { joinClasses } from "../helpers/dom"

export const CircleButton = ({
  children,
  onClick,
  isActive,
  title,
}: {
  children?: ReactNode
  onClick?: () => void
  isActive: boolean
  title?: string
}) => (
  <button
    className={joinClasses([
      "c-circle-button",
      isActive && "c-circle-button--active",
    ])}
    onClick={onClick}
    title={title}
    aria-pressed={isActive}
  >
    <div className="c-circle-button__content">{children}</div>
  </button>
)
