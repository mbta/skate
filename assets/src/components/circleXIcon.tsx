import React, { ComponentPropsWithoutRef } from "react"

/**
 * Interactive variant of the `CircleXIcon`.
 *
 * Implements focus and hover states via CSS and SVG.
 */
export const CircleXIcon = (props: ComponentPropsWithoutRef<"svg">) => (
  <svg
    className="c-circle-x-icon"
    viewBox="0 0 48 48"
    xmlns="http://www.w3.org/2000/svg"
    {...props}
  >
    <circle
      cx="50%"
      cy="50%"
      r="75%"
      className="c-circle-x-icon__focus-circle"
    />
    <circle cx="50%" cy="50%" r="75%" className="c-circle-x-icon__white-fill" />
    <path
      className="c-circle-x-icon__path"
      d="m24 .05a24 24 0 1 0 24 23.95 24 24 0 0 0 -24-23.95zm13.19 32.54a3.33 3.33 0 1 1 -4.73 4.69l-8.46-8.57-8.57 8.48a3.33 3.33 0 1 1 -4.69-4.73l8.55-8.46-8.48-8.57a3.33 3.33 0 1 1 4.73-4.69l8.46 8.55 8.57-8.48a3.33 3.33 0 0 1 2.35-1 3.35 3.35 0 0 1 3.33 3.35 3.33 3.33 0 0 1 -1 2.35l-8.54 8.49z"
    />
  </svg>
)
