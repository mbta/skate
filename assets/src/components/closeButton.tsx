import React from "react"
import { closeIcon } from "../helpers/icon"

interface Props {
  onClick: () => void
}

const CloseButton = ({ onClick }: Props) => (
  <button className="m-close-button" onClick={onClick}>
    {closeIcon("m-close-button__icon")}
  </button>
)

export default CloseButton
