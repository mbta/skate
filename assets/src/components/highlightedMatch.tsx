import React from "react"
import { intersperseString } from "../helpers/array"
import { filterToAlphanumeric } from "../models/searchQuery"

export const HighlightedMatch = ({
  content,
  highlightText,
}: {
  content: string
  highlightText?: string
}): JSX.Element => {
  if (highlightText === undefined) {
    return <>{content}</>
  }

  const match = content.match(highlightRegex(highlightText))

  if (match === null || match.index === undefined) {
    return <>{content}</>
  }

  const matchingString = match[0]

  return (
    <>
      {[
        content.slice(0, match.index),
        <span className="highlighted" key={`highlighted-${match.index}`}>
          {matchingString}
        </span>,
        <HighlightedMatch
          content={content.slice(match.index + match[0].length)}
          highlightText={highlightText}
          key={`highlighted-extension-${match.index + match[0].length}`}
        />,
      ]}
    </>
  )
}

const highlightRegex = (highlightText: string): RegExp => {
  const stripped = filterToAlphanumeric(highlightText)
  const allowNonAlphanumeric = intersperseString(stripped, "[^0-9a-zA-Z]*")
  return new RegExp(allowNonAlphanumeric, "i")
}
