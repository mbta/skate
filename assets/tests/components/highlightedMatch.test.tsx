import React from "react"
import renderer from "react-test-renderer"
import { HighlightedMatch } from "../../src/components/highlightedMatch"

describe("HighlightedMatch", () => {
  test("renders the content, wrapping the matching text in a span", () => {
    const content = "SMITH #20138713820"
    const highlightText = "138"

    const tree = renderer.create(
      <HighlightedMatch content={content} highlightText={highlightText} />
    )

    expect(tree).toMatchSnapshot()
  })

  test("is insensitive in its matching", () => {
    const content = "SMITH #201387 tmits"
    const highlightText = "mit"

    const tree = renderer.create(
      <HighlightedMatch content={content} highlightText={highlightText} />
    )

    expect(tree).toMatchSnapshot()
  })

  test("ignores spaces and hyphens", () => {
    const content = "abcde-f gh"
    const highlightText = "b c-defg"
    const tree = renderer.create(
      <HighlightedMatch content={content} highlightText={highlightText} />
    )

    expect(tree).toMatchSnapshot()
  })

  test("can highlight the whole string", () => {
    const content = "abc"
    const highlightText = "abc"
    const tree = renderer.create(
      <HighlightedMatch content={content} highlightText={highlightText} />
    )

    expect(tree).toMatchSnapshot()
  })

  test("renders the original content if no highlight text is specified", () => {
    const content = "SMITH #201387"

    const tree = renderer.create(<HighlightedMatch content={content} />)

    expect(tree).toMatchSnapshot()
  })
})
