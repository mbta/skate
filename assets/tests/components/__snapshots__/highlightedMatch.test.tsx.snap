// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightedMatch can highlight the whole string 1`] = `
<span
  className="highlighted"
>
  abc
</span>
`;

exports[`HighlightedMatch ignores spaces and hyphens 1`] = `
[
  "a",
  <span
    className="highlighted"
  >
    bcde-f g
  </span>,
  "h",
]
`;

exports[`HighlightedMatch is insensitive in its matching 1`] = `
[
  "S",
  <span
    className="highlighted"
  >
    MIT
  </span>,
  "H #201387 t",
  <span
    className="highlighted"
  >
    mit
  </span>,
  "s",
]
`;

exports[`HighlightedMatch renders the content, wrapping the matching text in a span 1`] = `
[
  "SMITH #20",
  <span
    className="highlighted"
  >
    138
  </span>,
  "7",
  <span
    className="highlighted"
  >
    138
  </span>,
  "20",
]
`;

exports[`HighlightedMatch renders the original content if no highlight text is specified 1`] = `"SMITH #201387"`;
